census/vanilla-tf/trainer/task.py

```diff
@@ -0,0 +1,349 @@
+  return accuracy
+
+#
+# Read the input and process
+#
+def get_inputs_and_labels():
+  """Placeholder for inputs and labels
+     input shape = [None, 364] label shape = [None, 2]."""
+  inputs = tf.placeholder(tf.float32, shape=[None, 346])
+  labels = tf.placeholder(tf.float32, shape=[None, 2])
+  return (inputs, labels)
+
+def read_input_data(file_name, skiprows=None):
+  """Read the input data as a pandas DataFrame of features and labels."""
+  input_df = pd.read_csv(StringIO(tf.gfile.GFile(file_name).read()), names=CSV_COLUMNS, skiprows=skiprows)
+
+  label_df = input_df.pop(LABEL_COL)
+  return (input_df, label_df)
+
+def generate_input(input_df, label_df):
+  """Prepare the input columns using SparseTensor."""
+
+  # convert the continuous columns into tf.constant tensor
+  continuous_columns = [
+      tf.constant(input_df[col].values) for col in CONTINUOUS_COLS
+  ]
+
+  # convert the categorical columns into sparse tensors
+  categorical_columns = [
+      tf.SparseTensor(
+          indices=[[i, 0] for i in range(input_df[col].size)],
+          values=input_df[col].astype('category').cat.codes.values,
+          dense_shape=[input_df[col].size, 1])
+      for col in CATEGORICAL_COLS
+  ]
+
+  # convert the labels into one hot encoding
+  label_tensor = tf.one_hot(
+      label_df.astype('category').cat.codes.values,
+      2, off_value=1, on_value=0)
+
+  return (
+      continuous_columns + categorical_columns, label_tensor
+  )
+
+
+#
+# Feature crosses and generation of wide columns
+#
+def sparse_cross(feature_tensors, num_buckets, name='cross'):
+  """Sparse feature cross of the feature SparseTensors."""
+  return sparse_feature_cross_op.sparse_feature_cross(
+      feature_tensors,
+      hashed_output=True,
+      num_buckets=num_buckets,
+      hash_key=tf.contrib.layers.SPARSE_FEATURE_CROSS_DEFAULT_HASH_KEY,
+      name=name
+  )
+
+def generate_wide_columns(input_columns):
+  """Generate wide columns by adding feature crosses of SparseTensors."""
+  (age, education_num, capital_gain, capital_loss, hours_per_week,
+   gender, race, education, marital_status, relationship, workclass,
+   occupation, native_country) = input_columns
+
+  # bucketize the age feature
+  # See https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/layers/python/ops/bucketization_op.py
+  age_bucket = bucketization_op.bucketize(age, [18, 25, 30, 35, 40, 45, 50, 55, 60, 65])
+
+  wide_columns = [
+      sparse_cross([education, occupation], 15 * 16, name='edu_occ'),
+      sparse_cross(
+        [age_bucket, race, occupation],
+        10 * 5 * 15,
+        name='age_race_occ'),
+      sparse_cross(
+        [native_country, occupation],
+        42 * 15,
+        name='native_country_occ'),
+      gender,
+      native_country,
+      education,
+      occupation,
+      workclass,
+      marital_status,
+      relationship,
+      age_bucket
+  ]
+
+  return wide_columns
+
+def concat_wide_columns(wide_columns):
```

Comment:
I think this method should be merged into `generate_wide_columns`.

Pull out all the vocab sizes into a module level variable `SPARSE_BUCKET_SIZES`, and then use a list comprehension to make the return value.